feat(example): friendly organizer

Chaegnal opened the pull request:
This realms is directed toward friend group that want to organize meet-up more easily. Each group has admins who can create DAO poll in order to execute certain action while ensuring the agreement of the majority. (r/chaegnal/organizer)

Uses the pure package 'p/nt/commondao'  for the DAO implementation.

Each file is tested and comments were added to enhance readability.

Chaegnal committed:
feat(dao): implements custom DAO

Chaegnal committed:
Merge branch 'gnolang:master' into master

Chaegnal committed:
feat(organizer): implements organizer realm

Chaegnal committed:
test(organizer): adds the first test for the organizer realm

Chaegnal committed:
Merge branch 'master' of github.com:Chaegnal/gno

Chaegnal committed:
fix(dao): removes redundant check

Chaegnal committed:
refacto(group): remove redundant function

Chaegnal committed:
refacto(organizer): exports function

Chaegnal committed:
test(organizer): testing of file

Chaegnal committed:
Merge branch 'gnolang:master' into master

Chaegnal committed:
test(group): testing of file

Chaegnal committed:
test(dao): testing of file

Chaegnal committed:
feat(dao): implements function to set hangout

Chaegnal committed:
fix(group): protects action to owner only and allows owner to give out ownership and close proposal prematurely

Gno2D2 commented:
#### 🛠 PR Checks Summary
 🔴 Pending initial approval by a review team member, or review from tech-staff


##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)
- [ ] The pull request description provides enough details


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🔴 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: Chaegnal/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🔴 At least 1 user(s) of the organization reviewed the pull request (with state "APPROVED")
    │       ├── 🔴 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🟢 This pull request is a draft
    └── 🔴 Then
        └── 🔴 And
            ├── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)
            └── 🔴 At least 1 user(s) of the team tech-staff reviewed pull request

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)
    └── 🟢 Not (🔴 Pull request author is user: dependabot[bot])

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

Chaegnal committed:
fix: changes the logic of the login process in order to go through query so the visual is not changed for everyone due to a global var

leohhhn commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,464 @@
+}
+
+// Verify that a user is in the whitelist and possesses the admin rank
+func (dao *DAO) IsAdmin(adr std.Address) bool {
+	isAdmin := false
+
+	member, isMember := dao.IsMember(adr)
+	if !isMember {
+		return false
+	}
+
+	if member.Rank == adminRank {
+		isAdmin = true
+	}
+
+	return isAdmin
+}
```
this can be simplified 

```suggestion
func (dao *DAO) IsAdmin(adr std.Address) bool {
	member, isMember := dao.IsMember(adr)
	return isMember && member.Rank == adminRank
}
```

Chaegnal replied:
[aac92fb](https://github.com/gnolang/gno/pull/4185/commits/aac92fb239790bd9f00bc64078049f419bf0366b) <- resolved in this commit ! Thanks :)

leohhhn commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,464 @@
+}
+
+// Verify if a user is in the waitlist
+func (dao *DAO) IsPending(adr std.Address) bool {
+	_, ok := dao.AwaitingApproval.Get(adr.String())
+	if !ok {
+		return false
+	}
+	return true
+}
+
```
can be simplified

```suggestion
func (dao *DAO) IsPending(adr std.Address) bool {
	_, ok := dao.AwaitingApproval.Get(adr.String())
	return ok
}

```

Chaegnal replied:
[9847f28](https://github.com/gnolang/gno/pull/4185/commits/9847f286bd776c5765a5770da7f2ef1e2707fab0) <- resolved in this commit !

leohhhn commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,464 @@
+}
+
+// Getters
+func (dao *DAO) GetWhitelisted() *avl.Tree {
+	return dao.Whitelist
+}
+
+func (dao *DAO) GetAwaiting() *avl.Tree {
+	return dao.AwaitingApproval
+}
+
+func (dao *DAO) GetProposals() *avl.Tree {
+	return dao.Proposals
+}
+
+func (dao *DAO) GetHangouts() *avl.Tree {
+	return dao.Hangouts
+}
```
If these trees have sensitive data, you should watch out how you expose them. check our `p/demo/avl/rotree` and see if it fits your usecase.

Chaegnal replied:
I used read-only tree for all getter except for the proposal one as it did not fit my use case for this tree in particular. Fixed in this commit: [3d6aebc](https://github.com/gnolang/gno/pull/4185/commits/3d6aebc3aaf21ab6b35cbcf2808c3b411cea1e05)

leohhhn commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,464 @@
+
+// Set user into the waitlist
+// Verify if the user is already in the waitlist or in the whitelist
+func (dao *DAO) PendingApproval(adr std.Address) {
+	_, err := dao.AwaitingApproval.Get(adr.String())
+	if err {
+		panic("Already awaiting approval.")
+	}
+
+	_, err = dao.IsMember(adr)
+	if err {
+		panic("Already whitelisted.")
+	}
+
+	dao.AwaitingApproval.Set(adr.String(), true)
+}
```
```suggestion
func (dao *DAO) PendingApproval(adr std.Address) {
	if _, exists := dao.AwaitingApproval.Get(adr.String()); exists {
		panic("Already awaiting approval.")
	}

	if _, isMember := dao.IsMember(adr); isMember {
		panic("Already whitelisted.")
	}

	dao.AwaitingApproval.Set(adr.String(), true)
}
```

leohhhn replied:
I don't think you should check if adr is a member here. This function should specifically check if there is a pending approval; then, in the function to actually start the approval process you panic if the person is already a member. This way there is no reason to check if the person is a pending member in this function, it can just never happen

leohhhn commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,464 @@
+	return true
+}
+
+func (dao *DAO) countAdmins() int {
```
This is a helper so it can be moved lower in the file for readability

Chaegnal replied:
Thanks ! -> [7ea8ec2](https://github.com/gnolang/gno/pull/4185/commits/7ea8ec220280670d287230fe136de7f9128336f0)

leohhhn commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,464 @@
+}
+
+// Move user from the waitlist to the whitelist
+// Verify that the action is done by an admin
+// Verify that the user is in the waitlist and not already in the whitelist
```
These are obvious if you read the code

Chaegnal replied:
I did not know how specific the comments needed to be, do you want me to remove all the comments that talk about the verifications of the functions ?

leohhhn replied:
This is just general advice; if you can read the code easily and understand what it does, there is no need for comments about it. this is especially true if you use proper variable naming and you write clean code. It's also something you get the feeling for when writing more and more code. No worries

leohhhn commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,464 @@
+func (dao *DAO) VoteProposal(voteId, choiceIdx int, caller std.Address) {
+	_, ok := dao.IsMember(caller)
+	if !ok {
+		panic("You are not whitelisted.")
```
```suggestion
		panic("you are not a part of this DAO")
```

Usually panics & error msgs in go/gno are lowercase without punctuation

Chaegnal replied:
Corrected each panics in this commit: [9f5a519](https://github.com/gnolang/gno/pull/4185/commits/9f5a519d2c30dcced3296dd149303c96fa659e55)

leohhhn commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,464 @@
+	dao.AwaitingApproval.Set(adr.String(), true)
+}
+
+// Move user from the waitlist to the whitelist
+// Verify that the action is done by an admin
+// Verify that the user is in the waitlist and not already in the whitelist
+func (dao *DAO) WhitelistingAwaitingApproval(caller, adr std.Address) {
+	if !dao.IsAdmin(caller) {
+		panic("Only admins can whitelist members.")
+	}
+
+	_, ok := dao.AwaitingApproval.Get(adr.String())
+	if !ok {
+		panic("This address is not awaiting approval.")
+	}
+
+	_, isMember := dao.IsMember(adr)
+	if isMember {
+		panic("Already whitelisted.")
+	}
+
+	dao.WhitelistMember(adr)
+	dao.AwaitingApproval.Remove(adr.String())
+}
```
Better name would just be "AcceptUser" or similar, no? It's a bit of a mouthful with wait/whitelist.
In the end, its just one waiting list and then you go into the DAO (become a member), right?

Chaegnal replied:
You are definitely right ! Changed the name to "WhitelistUser" in this commit: [8a4827f](https://github.com/gnolang/gno/pull/4185/commits/8a4827f9f6f33fcc34f24dfc27ce62579de6bb0f)

leohhhn commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,464 @@
+}
+
+// Set a user into the whitelist with a default member rank
+func (dao *DAO) WhitelistMember(adr std.Address) {
+	newWhitelisted := WhitelistedUser{
+		Address: adr,
+		Rank:    memberRank,
+	}
+
+	dao.Whitelist.Set(adr.String(), newWhitelisted)
+}
```
Not sure i get the point of this since you have this whole system of whitelisting with approvals - this function is not protected so anyone can whitelist anyone with a default rank? Or I'm missing something

Chaegnal replied:
I need it to set the creator of the group as a member itself as there are no admin yet, but you are right it should be protected! Here it goes: [40166b4](https://github.com/gnolang/gno/pull/4185/commits/40166b45098133227bc8ccb17d7901ffd7d6e4bc)

leohhhn commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,464 @@
+
+// Set a default admin at the creation of the DAO
+// Private function to maximize security
+func (dao *DAO) defaultAdmin(adr std.Address) {
```
Helper, move it down the file

Chaegnal replied:
Done, also renamed the whitelistMember function to show that the lack of admin check: [7c5cbac](https://github.com/gnolang/gno/pull/4185/commits/7c5cbac6af478e154c1c434734d3ac16b2b89d23)

leohhhn reviewed (commented):
Leaving some comments on the DAO package - please, try to split up the file into smaller pieces based on some topic or similar, to make it easier to understand and read the code. This applies for all code you wrote :)

I'll take another look when you fix the current comments

Chaegnal commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,464 @@
+	Title        string
+	Description  string
+	Owner        std.Address
+	Callback     string
+	CallbackArgs []string
```
Is it possible to use function reference in Gno ? If yes then i could implement that and remove the switch case that match the string to a supported function. Then at the creation of the proposal, user would have to directly pass a function, is that right ?

Chaegnal committed:
refacto(dao): simplify IsPending function

Chaegnal committed:
refacto(dao): simplify IsMember function

Chaegnal committed:
fix: uses read only tree for most of the getter to protect data

Chaegnal committed:
refacto(dao): move helper function to count admin in the DAO to the end of the file

Chaegnal committed:
refacto: remove capitalization and ponctuaction in panic messages

Chaegnal committed:
refacto: simplify the name of the fuction to whitelist users

Chaegnal committed:
fix(dao): protects function to whitelist user without admin check

Chaegnal committed:
refacto(dao): moved helper function at the end of the file

Chaegnal committed:
refacto: splitting up the base code into smaller more comprehensive files

Chaegnal commented:
> Leaving some comments on the DAO package - please, try to split up the file into smaller pieces based on some topic or similar, to make it easier to understand and read the code. This applies for all code you wrote :)
> 
> I'll take another look when you fix the current comments

In this commit: [d5ca4e8](https://github.com/gnolang/gno/pull/4185/commits/d5ca4e84b1530c7a5db74496c00ffe2b8001c95f)

I split up the code into smaller, more comprehensive files. I could also, in the `r/chaegnal/organizer`, split the types in one separate file, let me know ! 

I will also redo the test in a future commit !

Chaegnal committed:
refacto(group): refactor query parser to take a keyword and reduce redundancy

Chaegnal committed:
refacto(group): move AdminCloseProposal to the end of the file to add more coherence

Chaegnal committed:
test(dao): fixes test and put them in coherent files

Chaegnal committed:
test(organizer): fixes the test of the organizer file

Chaegnal committed:
test(group): fixes test for the group page

Chaegnal committed:
Merge pull request #3 from Chaegnal/test/fix-test

Test/fix test

Chaegnal committed:
Merge branch 'gnolang:master' into fix/implement-cross

Chaegnal committed:
Merge branch 'fix/implement-cross' of github.com:Chaegnal/gno into fix/implement-cross

Chaegnal committed:
style(organizer): run make tidy

Chaegnal commented:
> Hi @Chaegnal . The CI check for mod-tidy says "Some gno.mod files are not tidy, please run 'make tidy'." Are you able to do this?

Hi @jefft0, I ran the make tidy ! The result of that can be found in this commit: [95a904f](https://github.com/gnolang/gno/pull/4185/commits/95a904fded2c3e65b0f392e9a180bc7bf70adc55)

Chaegnal committed:
chore(organizer): run make tidy

Chaegnal committed:
fix(organizer): implement crossing

Chaegnal committed:
Merge pull request #4 from Chaegnal/fix/implement-cross

Fix/implement cross

Chaegnal committed:
Merge branch 'gnolang:master' into master

Chaegnal committed:
test(group): update test to fit crossing, failing

Chaegnal committed:
test(organizer): fixes the test that needed the interrealm

Chaegnal committed:
Merge pull request #5 from Chaegnal/fix/implement-cross

Fix/implement cross

Chaegnal committed:
test(organizer): enhance clarity by using uassert

Chaegnal committed:
test(dao): enhance clarity by using uassert

leohhhn commented on examples/gno.land/r/chaegnal/organizer/organizer.gno:
```diff
@@ -0,0 +1,133 @@
+
+// Helpers
+
+// Return the user's registered name or address
+func ResolveAddressName(addr std.Address) string {
+	user := users.ResolveAddress(addr)
+	if user != nil {
+		return user.Name()
+	}
+	return addr.String()
+}
```
nitpicking here, but: not sure why you'd want this function, or maybe the naming is a bit off.

Why not, in the calling code, just do:
```go
user:= users.ResolveAddress(addr)
if user != nil {
	renderString += user.Name()
} else {
	renderString += addr.String()
}
```

The func giving bad vibes because the function is doing two things depending on the input: returning a text name, or just converting from one type to the other.

leohhhn reviewed (commented):
Leaving some more comments

Chaegnal committed:
Merge branch 'gnolang:master' into master

Chaegnal committed:
refacto(organizer): use std.PreviousRealm instead of std.OriginCaller

Chaegnal committed:
refacto(organizer): renamed txSender var to caller for more consistancy

Chaegnal committed:
doc: make the documentation gnodoc compliant

Chaegnal committed:
chore: renaming ResolveAdressName to NameOrAddress to enhance readability

Chaegnal committed:
Merge pull request #6 from Chaegnal/refacto/comments

Refacto/comments

Chaegnal commented on examples/gno.land/r/chaegnal/organizer/organizer.gno:
```diff
@@ -0,0 +1,133 @@
+package organizer
```
Added some documentation and made it godoc compliant here -> [5c9c866](https://github.com/gnolang/gno/pull/4185/commits/5c9c866fe28998424cb2882a94280615f71f6bb8)

Chaegnal commented on examples/gno.land/p/chaegnal/dao/dao.gno:
```diff
@@ -0,0 +1,30 @@
+	Hangouts         *avl.Tree
+}
+
+// Create a new DAO
+// Initializes proposal, whitelist and waitlist
+func NewDAO(caller std.Address) *DAO {
```
Solved this here -> [5c9c866](https://github.com/gnolang/gno/pull/4185/commits/5c9c866fe28998424cb2882a94280615f71f6bb8)

Chaegnal commented on examples/gno.land/r/chaegnal/organizer/organizer.gno:
```diff
@@ -0,0 +1,133 @@
+// Triggered when a user request to join a group
+// Put the user in a waitlist until an admin let them join
+// Verify if the group exist and if the user is already a member
+func AskJoinFriendGroup(groupId int) {
```
Solved here, thanks for showing me -> [5c9c866](https://github.com/gnolang/gno/pull/4185/commits/5c9c866fe28998424cb2882a94280615f71f6bb8)

Chaegnal commented on examples/gno.land/r/chaegnal/organizer/organizer.gno:
```diff
@@ -0,0 +1,133 @@
+// Set the creator as the default admin
+func CreateFriendGroup(title string) {
+	crossing()
+	caller := std.OriginCaller()
```
Switched OriginCaller to PreviousRealm here -> [88fffc0](https://github.com/gnolang/gno/pull/4185/commits/88fffc0846cb8b6e37423d02883653c80ad2519a)

Renamed the txSender var into caller here -> [2d0d4d9](https://github.com/gnolang/gno/pull/4185/commits/2d0d4d99633e896c0579b19ec20d2ae60bb8e7bf)

Chaegnal committed:
feat(organizer): uses commondao for member management

Chaegnal committed:
fix(organizer): set the creator of the group as its first memmber and give admin rank

Chaegnal committed:
fix(organizer): fixes the displaying of proposals, ranking iota and group creation

leohhhn commented on examples/gno.land/p/chaegnal/dao/dao.gno:
Did you have a chance to update this?

Chaegnal committed:
Merge branch 'gnolang:master' into feat/integrate-common-dao

Chaegnal committed:
feat: start proposal implementation of commonDao

Chaegnal committed:
Merge branch 'gnolang:master' into fix/interreal-update

Chaegnal committed:
fix: implement the new syntax of interrealm

Chaegnal committed:
Merge pull request #7 from Chaegnal/fix/interreal-update

Fix/interreal update

Chaegnal committed:
Merge branch 'gnolang:master' into feat/integrate-common-dao

Chaegnal committed:
fx: removes the unnecessary cur argument in txcall

Chaegnal committed:
feat: display hangout

Chaegnal committed:
feat(organizer): finishes integrating commonDAO

Chaegnal committed:
chore(organizer): remove custom dao pure package

Chaegnal committed:
chore: removal of useless struct

Chaegnal committed:
refacto(organizer): moves hangout logic into separate file

Chaegnal committed:
doc(organizer): updates gnodoc

Chaegnal committed:
doc(organizer): updates gnodoc

Chaegnal committed:
refacto(organizer): adds iota instead of incrementing constants

Chaegnal committed:
Merge branch 'gnolang:master' into feat/integrate-common-dao

Chaegnal committed:
Merge pull request #8 from Chaegnal/feat/integrate-common-dao

Feat/integrate common dao

Chaegnal committed:
Merge branch 'gnolang:master' into master

Chaegnal committed:
fix(organizer): changes gno.mod to gnomod.toml

Chaegnal committed:
test(organizer): fixes test for cross function

Chaegnal committed:
test(organizer): fixes test after commonDAO integration

Chaegnal committed:
chore: fixes gnodoc

Chaegnal committed:
Merge branch 'gnolang:master' into test/commondao-integration

Chaegnal committed:
test(organizer): fixes test to fit commonDAO

Chaegnal committed:
Merge pull request #9 from Chaegnal/test/commondao-integration

Test/commondao integration

Chaegnal committed:
Merge branch 'gnolang:master' into master

Chaegnal committed:
chore(organizer): uses aborts uassert function for testing crossing func

jeronimoalbi commented on examples/gno.land/r/chaegnal/organizer/hangout.gno:
```diff
@@ -0,0 +1,29 @@
+// SetHangoutTimeAndPlace creates a hangout after validation from the dao
+// Set the new hangout into a dedicated AVL Tree
+func (g *FriendGroup) SetHangoutTimeAndPlace(date, time, place, desc string) error {
+	id := g.Hangout.Size() + 1
+	hangout := HangoutInfo{
+		Id:          id,
+		Date:        date,
+		Time:        time,
+		Place:       place,
+		Description: desc,
+		Active:      true,
+	}
+	g.Hangout.Set(strconv.Itoa(id), hangout)
```
You could consider using `gno.land/p/demo/seqid` package, or an approach similar to the one used in that package, to generate incremental IDs and also the corresponding AVL keys, to make sure AVL tree items are properly sorted so they can be iterated keeping lexicographical order, otherwise it would for example happen that ID "10" would be iterated or displayed before ID "2".

Chaegnal replied:
Hi Jeronimo,

I did the change to use sequid in this commit: [0f3baef](https://github.com/gnolang/gno/pull/4185/commits/0f3baef2e4c4e62e4afa66c45274c3a32a1b0968)

I was not sure what you mean about the lexicographical order, so i used the ID in form of a string. Let me know if that works for you.

Chaegnal committed:
refacto(organizer): use sequid for identification of hangouts and groups

Chaegnal committed:
Merge pull request #10 from Chaegnal/refacto/use-sequid

refacto(organizer): use sequid for identification of hangouts and groups

jeronimoalbi commented on examples/gno.land/r/chaegnal/organizer/organizer.gno:
```diff
@@ -0,0 +1,154 @@
+		Hangout:  avl.NewTree(),
+	}
+
+	newGroup.DAO.Members().Add(caller)
```
You could remove this, because `commondao.WithMember(caller)` already adds the caller as a member when the DAO is created:

```suggestion
```

Chaegnal replied:
Same thing, i was not sure but i removed it here -> [fc495a6](https://github.com/gnolang/gno/pull/4185/commits/fc495a6145a5b2353f447a296c7146fc9058a4bb)

jeronimoalbi commented on examples/gno.land/r/chaegnal/organizer/organizer.gno:
```diff
@@ -0,0 +1,154 @@
+func AskJoinFriendGroup(cur realm, groupId int) {
+	caller := std.PreviousRealm().Address()
+
+	raw, exists := ListGroups.Groups.Get(strconv.Itoa(groupId))
```
You would need to cast the `int` groupID into `seqid.ID` doing something like this to make sure the ID as string is right:

```suggestion
	id := seqid.ID(groupId).String()
	raw, exists := ListGroups.Groups.Get(id)
```

Chaegnal replied:
[ab3fd45](https://github.com/gnolang/gno/pull/4185/commits/ab3fd45db2b6e5edf88b7da913c9ab37fea04874)

jeronimoalbi commented on examples/gno.land/r/chaegnal/organizer/render.gno:
```diff
@@ -0,0 +1,180 @@
+	var out string
+
+	if adr != "" {
+		memberSet := commondao.NewMemberSet(group.DAO.Members())
+
+		if memberSet.Has(adr) {
```
This would also work:

```suggestion
		if group.DAO.Members().Has(adr) {
```

Chaegnal committed:
refacto(organizer): privatize nextHangoutID

Chaegnal committed:
refacto(orgnanizer): remove reassignation of sequid

Chaegnal committed:
refacto(organizer): remove useless addition of member at creation

Chaegnal committed:
refacto(organizer): changes the ID from a string to a seqid forthe hangout and friendgroup struct

Chaegnal committed:
refacto(organizer): discard the memberset to verify if the group has a member

Chaegnal committed:
refacto(organizer): discard the memberset to verify if the group has a member in render

Chaegnal committed:
test(organizer): fix test impacted by change to seqid

Chaegnal committed:
Merge pull request #11 from Chaegnal/refacto/seqid

Refacto/seqid

Chaegnal commented on examples/gno.land/r/chaegnal/organizer/organizer.gno:
```diff
@@ -0,0 +1,154 @@
+	group := raw.(FriendGroup)
+	memberSet := commondao.NewMemberSet(group.DAO.Members())
+
+	if memberSet.Has(caller) {
```
Did it here -> [8e03cac](https://github.com/gnolang/gno/pull/4185/commits/8e03cacbca7b2dc34ad6ac76cfd692c1b35c8c66)

Chaegnal commented on examples/gno.land/r/chaegnal/organizer/organizer_test.gno:
Indeed, I found quite troubling that the unit test carried on the state to other test. I'm going to try using filetest by taking example on what you did for commonDao realm, I'll let you know if I have issues :+1:

jeronimoalbi replied:
Filetests are quite open on how you name files, or how you write the test case (or cases) within each `*_filetest.gno` file.

Until now, as far as I know, by convention the name of the files start with "z_" so they the files are listed at the end, after the Gno code. Followed by a combination of a number and a letter, also by convention. For example, if you think that each filetest file is a test case:
- z_a_0_filetest.gno: Case 0 for a test
- z_a_1_filetest.gno: Case 1 for the same test
- z_b_0_filetest.gno: Case 0 for another test

Other approach could be to use the name of the public realm function you are testing as prefix:
- z_MyPublicFunction_0_filetest.gno
- z_MyPublicFunction_1_filetest.gno
- z_OtherPublicFunction_0_filetest.gno

Or if you want to remove the "z_" prefix from the name you could use a sub-directory named for example:
- `organizer/filetests`: Filetests for `organizer` realm
- `organizer/group/filetests`: Filetests for `group` realm

Test in these sub-directory will be run when you call something like:
```gno test ./...```

Chaegnal committed:
test(organizer): changes the test approach to filetest instead of unit test for public functions

jeronimoalbi commented on examples/gno.land/r/chaegnal/organizer/organizer.gno:
```diff
@@ -0,0 +1,151 @@
+	"gno.land/r/sys/users"
+)
+
+const (
+	VoteYay = 1 // VoteYay represented yes votes
+	VoteNay = 0 // VoteNay represented no votes
+
+	MemberRank = iota // memberRank represents the basic right of user
+	AdminRank         // adminRank represents the rank of admin with heightened rights
+)
```
You might want to split these into two if the intention is that `MemberRank` value should be 0:

```suggestion
const (
	VoteYay = 1 // VoteYay represented yes votes
	VoteNay = 0 // VoteNay represented no votes
)

const (
	MemberRank = iota // memberRank represents the basic right of user
	AdminRank         // adminRank represents the rank of admin with heightened rights
)
```

Right now `MemberRank == 2` and `AdminRank == 3`. The reason for that is that within the same `const` the `iota` in this case starts with 2. Splitting into independend `const` groups makes `iota` start with 0.

Chaegnal replied:
Thanks, it doesnt change how it works but it will indeed make more sense ! Here is the change: [4a28527](https://github.com/gnolang/gno/pull/4185/commits/4a285275888be5c9749726dc3b185163aae2593b)

Chaegnal committed:
fix(orgnaizer): fixes iota value

jeronimoalbi commented on examples/gno.land/r/chaegnal/organizer/group/member_actions.gno:
```diff
@@ -0,0 +1,102 @@
+)
+
+// AcceptAwaitingMember accepts new members into friendgroup and DAO
+func AcceptAwaitingMember(cur realm, groupId, logged string, adr std.Address) error {
+	if !isAdmin(groupId, std.Address(logged)) {
+		panic("only admin can do this action")
+	}
```
I believe you should consider the caller of the `AcceptAwaitingMember()` function as the logged:

```suggestion
func AcceptAwaitingMember(cur realm, groupId, adr std.Address) error {
	logged := std.PreviousRealm().Address()
	if !isAdmin(groupId, logged) {
		panic("only admin can do this action")
	}
```

Otherwise anyone could send the admin address argument when calling the function. By using `std.PreviousRealm().Address()` you make sure that only the admin can call this function.

Chaegnal replied:
I agree for some functions but im wondering about the functions, like AcceptAwaitingMember, that can be a callback. Because I check if the execute function everytime someone votes, that means that if the last voter is not an admin the execution of the proposal will fail.

A work-around for this issue would be to not check if the proposal is ready to be executed with each vote but via an interaction from an admin. Thus the std.PreviousRealm().Address() would be of an admin everytime. What do you think about it ?

jeronimoalbi replied:
The current issue with `AcceptAwaitingMember`, and other callbacks, is that they are public realm functions, that means anyone could directly call them, for example `AcceptAwaitingMember`, and send any arguments, so if one knows the admin address then it can use it to accept members skipping the proposal.

Nothing stops users from creating a realm or script to call the callbacks without proposals, for example:
```go
import "gno.land/r/chaegnal/organizer/group"

func init() {
  admin := "... valid admin address ..." // <- anyone knowing this address would be a security issue for the realm
  group.AcceptAwaitingMember(cross, "1", admin, "... invited user address ...")
}
```

If you still want to send the admin/logger address as argument you would have to make callbacks private so only `group` realm can call them

Chaegnal committed:
fix(organizer): check if address retrieved from the query parameter is valid

Chaegnal commented on examples/gno.land/r/chaegnal/organizer/render.gno:
```diff
@@ -0,0 +1,177 @@
+func displayJoiningStatus(group FriendGroup, adr std.Address) string {
+	var out string
+
+	if adr != "" {
```
You are right, i did it here -> [3b70b98](https://github.com/gnolang/gno/pull/4185/commits/3b70b98cc51448e13f620545ba7fb024e39245a7)

Chaegnal committed:
test(organizer): uses testutils to generate valid adresses

leohhhn committed:
Merge branch 'master' into master

jeronimoalbi commented on examples/gno.land/r/chaegnal/organizer/render.gno:
```diff
@@ -0,0 +1,177 @@
+	out += md.H2(md.Link("Create New FriendGroup", txlink.Call("CreateFriendGroup", "title", "default")) + "\n\n")
+	out += md.HorizontalRule()
+
+	addressStr := parseQuery(rawURL)
```
Just in case, if you choose to use `mux` package you can see a usage example in [commondao realm's render](https://github.com/gnolang/gno/blob/2971303b6c8afe1af586586945c3cad4dd3a764c/examples/gno.land/r/nt/commondao/render.gno#L22)

@Chaegnal feel free to resolve the review comment if you won't use `mux` or `realmpath`

jeronimoalbi commented on examples/gno.land/r/chaegnal/organizer/z_1_a_filetest.gno:
```diff
@@ -0,0 +1,39 @@
+)
+
+func init() {
+	testing.SetOriginCaller("Chaegnal")
```
Realm functions can be directly called either by another realm or by the user that is sending the transaction. When you refer to origin caller you are referring to the user sending the transaction.

> Indeed, i kept on using SetOriginCaller(), should i switch to the other ? What is the prefered way to do it ?

It will depend on your realm logic, and if you use `std.PreviousRealm().Address()` or `std.OriginCaller()` in your realm function.

I believe the preferred way is generally to use `std.PreviousRealm()` whenever possible, which means in test you would use something like:
```go
testing.SetRealm(std.NewUserRealm(user))
```

You can also check if the previous realm is a user when your logic requires it, when the ideas of a public function is to be called by a user and not by another realm:
```go
if std.PreviousRealm().IsUser() {
  // ...
}
```

Chaegnal committed:
test(organizer): fixes the test and uses the SetRealm() instead or OriginCaller()

Chaegnal committed:
fix(organizer): fixes security issue by privatizing proposal struct

Chaegnal committed:
fix(organizer): fixes empty tables and badly named titles

Chaegnal committed:
fix(organizer): fixes security issues by making functions private

Chaegnal committed:
test(organizer): fixes the test and rename the filetest

Chaegnal committed:
Merge branch 'gnolang:master' into master

jeronimoalbi commented on examples/gno.land/r/chaegnal/organizer/group/group_actions.gno:
```diff
@@ -0,0 +1,39 @@
+
+// DeleteFriendGroup allows the owner to delete the friendgroup
+// Panics if anyone but the owner try to delete the group
+func DeleteFriendGroup(cur realm, groupId, logged string) {
+	group := organizer.GetGroupFromId(groupId)
+
+	if group.Owner.Owner() != std.Address(logged) {
+		panic("only the owner can delete the group")
+	}
```
Recommendation for security, you should never trust input values, in this case using `std.PreviousRealm().Address()` is one way to make sure group owner address can be faked:
```suggestion
func DeleteFriendGroup(cur realm, groupId string) {
	logged := std.PreviousRealm().Address()
	group := organizer.GetGroupFromId(groupId)

	if group.Owner.Owner() != logged {
		panic("only the owner can delete the group")
	}
```

jeronimoalbi commented on examples/gno.land/r/chaegnal/organizer/group/proposal_actions.gno:
```diff
@@ -0,0 +1,155 @@
+
+	group := organizer.GetGroupFromId(groupId)
+
+	callbackArgsStr = strings.Trim(callbackArgsStr, "\"")
+	callbackArgs := strings.Split(callbackArgsStr, ",")
+
+	def := &callbackProposalDefinition{
+		GroupId:      groupId,
+		TitleValue:   title,
+		BodyValue:    desc,
+		Callback:     callback,
+		CallbackArgs: callbackArgs,
+	}
```
The proposal definition could store the caller (logged) address instead of receiving it though an arguments string, that way you can make sure is the right one and would remove redundancy

thehowl commented:
Hi @Chaegnal — thanks for the contribution! Closing as part of a stale-PR housekeeping sweep; this has been inactive since Jul 2025. Feel free to reopen or submit a fresh version if you'd like to continue.
